### ISSUE TITLE
[TASK] [Refactor] Unified Agent

### ISSUE THREAD

**Issue by morantr** [prplfoundation/prplMesh#1082]
## Description

The agent architecture is complicated and does not fit to MultiAP standard. 
It has the backhaul manager as an internal switch that forwarding the messages to multiple "son_slaves" which does the same things, but for different radios.
In MultiAp standard there are TLVs that should contain information on both radio, but since we have two separate entities for the radios, we need to send these messages twice.

### Benefits:

1. Reducing processes number:
The current number of processes:
Main Agent, Son_slave0, Son_slave2, Monitor0, Monitor2 - Total 6 Processes
The new architecture number of processes:
Agent, Fronthaul_manager0, Fronthaul_manager2 - Total 3 Processes
**Reduction in 3 processes.**

2. Reducing threads number:
The current number of threads:
Platform_manager, Backhaul_manager, Son_slave0, Son_slave2, Ap_manager0, Ap_manager2, Monitor0, Monitor2 - Total 8 threads
The new architecture number of threads:
Platform_manager, Backhaul_manager, Agent, Ap_manager0, Ap_manager2, Monitor0, Monitor2 - Total 7 threads
**Reduction in 1 thread.**

3. Backhaul manager simplicity.
Currently, the backhaul manager has responsibilities that it should not have, such as:
   - Messages forwarding to son_slaves.
   - Managing as topology manager
   - Searching for the controller with Autoconfig-search message.

   After Unified Agent implementation, the backhaul manager will only be responsible for a backhaul link establishment, and it will not have access to the bus.
  **The key benefit of it, is that there will no more message forwarding from the backhaul manager to two son slaves, and total fewer messages transmitted in the system**

4. One entity controlling the Agent.
Currently, there are two son_slaves that do the same things and send the same messages as duplicates for each radio entity.
On the new architecture, there will be only one Agent thread which will send messages containing information on both radios as the standard allows.
   - Fewer total messages will be sent.
   - Lesser prints to the logs.
   - Simpler handling in the controller side.

5. Lower memory consumption.
Ther are some memory fields that are duplicate for both son_slaves, which unified Agent will save.

### The downside of the new architecture:
Unless we unify the monitor and the ap_manager logs, we would have to create two more log files for the ap_manager threads.

![image](https://user-images.githubusercontent.com/10611984/78244237-dfdb8c00-74ed-11ea-9ccd-cce3e7f734b9.png)

### Phase 1 - Unify ap_manager and monitor thread under the same process.
- [x] Move the ap_manager thread to the monitor process.
- [x] Change the monitor name to "fronthaul".
- [x] Separate ap_manager and monitor logs to different files.
- [x] On fronthaul process, run the monitor only when the ap manager finished the attach sequence.
There is no need for both of them to do the attach at the same time, once one attach sequence is succeeded the other one will succeed too.
- [x] Replace the ap_manager thread start and monitor bring up by fronthaul process bring-up on the son_slave.
- [x] Add notification message from ap_manager to son_slave, notifying that te ap_manager is up.
- [x] Add configuration message from son_slave to ap_manager, containing configuration pulled from the platform manager, which will be sent to the ap_manager after receiving up-notification from it.

### Phase 2 - Create Agent database class.
Currently, a lot of the agent data is duplicated among several processes and threads.
To pass the data, we usually need to add some parameters to a message or add a new message.
Creation of a database class for the Agent which will contain all the information the son_slave is currently saving will make the system more flexible and will make some of the changes and flows movement to be more easily done, and in the future, it could be used for memory optimization or encapsulating some submodules (like topology class for example).
- [ ] Create an Agent database class.
   - [ ] Copy all data members on the backhaul manager not related to link establishment to the Agent database.

### Phase 3 - Create an Agent thread
- [ ] Create a skeleton for Agent class, which will contain some basic member variables and function from son_slave. The Agent will inherit from BTL.
- [ ] Use Agent DB - On each piece of code that are moved from son_slave, use Agent DB instead of son_slave class members. 
- [ ] Agent FSM: Connect and register to the platform manager.
   - [ ] The Agent will register with both radios.
   - [ ] The Platform manager will fill information on all radio and send a response.
   - [ ] The Platform can remove the work queue since it is no longer needed.

- [ ] Agent FSM: Bring up the Fronthauls. 
   - [ ] Bring up the Fronthaul processes.
   - [ ] Connect and register to the AP managers and the monitor threads.
   - [ ] Add mechanism that waits until both Fronthauls are up.

- [ ] Agent FSM: Connect and request for link establishment from the backhaul manager. On the son_slave we connect and register to the backhaul manager before we bring up the AP, and establish the backhaul link after it's up. Since now we have unified Agent there is no need for the backhaul register-enable mechanism anymore. The whole thing could be replaced with a single link establishment request.
   - [ ] Replace backhaul inheritance from BTL to socket_thread, and remove the message forwarding mechanism.
   - [ ] Backhaul manager slave_sockets refactor. Currently, the backhaul manager uses a son_slave sockets container which will not be necessary anymore. Need to refactor it so it will not use the son_slave socket.
    - [ ] Once the backhaul manager establishes the link, return a notification to the Agent and switch to the "Link established" state.

- [ ] Move Auto-Configuration search message flow from the backhaul manager to the Agent. (Need to send once for each band)
   - [ ] Add mechanism that waits for responses for the two search messages.
- [ ] Move UCC-Listener from the backhaul manager to the Agent.
- [ ] Agent FSM: Add Auto-Configuration-WSC
   - [ ] Move Auto-Configuration-WSC from son_slave to the Agent thread.
   - [ ] Send one message for each radio.
   - [ ] Add mechanism that waits for responses for the two Auto-Configuration-WSC messages.
- [ ] Move CMDU handlers from backhaul manager to the Agent:
   - [ ] TOPOLOGY_QUERY_MESSAGE
   - [ ] AP_CAPABILITY_QUERY_MESSAGE
   - [ ] HIGHER_LAYER_DATA_MESSAGE
   - [ ] LINK_METRIC_QUERY_MESSAGE
   - [ ] COMBINED_INFRASTRUCTURE_METRICS_MESSAGE
   - [ ] CLIENT_CAPABILITY_QUERY_MESSAGE
   - [ ] ASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE

- [ ] Move Topology Discovery from the backhaul manager to the Agent.
- [ ] All CMDUs which are handled on son_slave shall be handled on the Agent.
- [ ]  Unify CHANNEL_PREFERENCE_REPORT_MESSAGE which is sent separately on each son_slave to one CMDU with multiple TLVS.

### Phase 4 - Fixups
- [ ] Fix any errors discovered on the system flow or test scenarios while starting to use Unified Agent. 

### Phase 5 - Cleanups 
- [ ] Remove any code leftovers that are no longer needed not being used.

## Development Stages
Each Phase can be done in on or more separate PRs.
Phase 2 - Will be done as infrastructure PR without actually use it.
Phase 3 - This is a kind of big Phase and probably will be needed to split into several PRs. Since each code part can't exist without other so all the code will be compiled and pass all the test successfully there are two options:
1. Make only on PR for the whole phase.

   Pros:
   - All the changes will be done on the backhaul manager and son_slave files.

   Cons:
   - A lot of commits - hard to review.
   - It will take a lot of time to merge.
   - Rebase to master will be hell.

2. Make separate new files for backhaul manager and Agent which will not be used until the task end, and be switched on Phase 4.

   Pros:
   - Could be split to several PRs without break anything - Easy to review.
   - Less time to merge single PR.
   - Rebase to master is a piece of cake.

    Cons:
   - Harder to track changes in review on pieces of code that are copied from file to file.

**I prefer option 2.**

**Comment by arnout** [comment]
> Separate ap_manager and monitor logs to different files

Why is this needed? We currently have an agent log file and 2 log files per radio. With the new architecture, that would be an agent log file an 1 log file per radio - monitor and apmanager log to the same file. Is there any issue with that?

> Unless we unify the monitor and the ap_manager logs, we would have to create two more log files for the ap_manager threads.

No we don't. The logs of ap_manager would still go to agent_radio0.log, while the logs of the agent go to agent.log and the logs of monitor go to monitor_radio0.log. We *would* need to change a lot of check_log calls in test_flows.py, but maybe that's what you meant. However, after @Galvien's #1020, almost of of these check_log calls would be replaced by sniffer checks.

> Phase 2 - Create an Agent thread

I think it will actually be a lot easier to simply rename backhaul to agent, and then create a new backhaul manager thread and move the backhaul functionality there. 90% of the code in backhaul_manager is in fact agent code.

And I think you're missing phase3 where son_slave code is moved to the agent thread, and son_slave is removed. This phase3 is something that has to be done in a single shot, else we'd need to maintain two UDS connections to ap_manager (one from agent and another one from son_slave).

Some additional observations:
- phase 1 and phase 2/3 are IMO pretty much independent of each other, they could be done as two separate tasks, in any order.
- AFAICS there is no real reason to have a separate thread for the monitor. The only reason for it to be a separate thread is to be able to handle monitor events while we're blocked on an ap command. But I don't really think this is needed in practice. That said, it's a lot of work to merge the two threads, with limited benefit.

**Comment by tomereli** [comment]
> > Separate ap_manager and monitor logs to different files
> 
> Why is this needed? We currently have an agent log file and 2 log files per radio. With the new architecture, that would be an agent log file an 1 log file per radio - monitor and apmanager log to the same file. Is there any issue with that?

The issue is that the monitor contains repeating logs which will cause the radio agent log file to wrap-around, and be much harder to read.
The solution is to refactor the logging information in the monitor log - I tried to push for this approach but APM team rejected saying that all logs are important. So for now, the only solution is to separate into 2 log files.
